src/librustc/traits/util.rs

```diff
@@ -504,6 +506,32 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         };
         ty::Binder((trait_ref, sig.skip_binder().output()))
     }
+
+    pub fn impl_is_default(self, node_item_def_id: DefId) -> bool {
+        match self.hir.as_local_node_id(node_item_def_id) {
+            Some(node_id) => {
+                let item = self.hir.expect_item(node_id);
+                if let hir::ItemImpl(_, _, defaultness, ..) = item.node {
+                    defaultness.is_default()
+                } else {
+                    false
+                }
+            }
+            None => {
+                self.global_tcx()
+                    .sess
+                    .cstore
+                    .impl_defaultness(node_item_def_id)
+                    .is_default()
+            }
+        }
+    }
+}
+
+impl<'a, 'gcx, 'tcx> NodeItem<hir::Defaultness> {
+    pub fn is_final(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
```

Comment:
I would probably have put this as a method on TyCtxt instead, as that’s how pretty much all such queries are implemented.